### PR TITLE
NF: Support BIDS format

### PR DESCRIPTION
**Pull request by GuillaumeTh**
Add the BIDS format support in input of TractoFlow.

2 new parameters:

--bids: take a folder following the BIDS structure

--bids_config: take the corrected json file generated by ReadBIDS process if some informations are not present.

ReadBIDS process generate a json file with image informations for each subjects. If an information is not present in the BIDS jsons, the process will add a "todo" value at the missing key in the json file generated by ReadBIDS. After that, you can correcte the ReadBIDS json and give this json at --bids_config parameter instead using --bids

### CODE REVIEW

File: bin/readBIDS.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                dwi_revPE = dwi_PE+'-'
+            else:
+                dwi_revPE = dwi_PE[0]
+
+        # Find b0 for topup, take the first one
+        revb0_path = ''
+        for nfmap in fmaps:
+            if 'PhaseEncodingDirection' in nfmap.metadata and\
+               'IntendedFor' in nfmap.metadata:
+                fmap_PE = nfmap.metadata['PhaseEncodingDirection']
```
Comment (arnaudbore) [inline bin/readBIDS.py]: doesn't seem to work for me -> use nfmap.get_metadata() instead

**Comment by GuillaumeTh** [comment]
I'm not sure but maybe it's better to push the python script in the singularity/docker instead of the bin folder ? @jchoude @arnaudbore What do you think ?

**Review by arnaudbore — APPROVED** [review]

**Comment by GuillaumeTh** [comment]
@jchoude when you will have time, can you review this ?

**Comment by arnaudbore** [comment]
@mdesco can you test it ?
@boiteaclou can you review the code ?

File: main.nf [hunk 2]
```diff
@@ -157,42 +161,135 @@ workflow.onComplete {
...
+    jsonSlurper = new JsonSlurper()
+        data = jsonSlurper.parseText(it.getText())
+        for (item in data){
+            sid = "sub-" + item.subject + "_ses-" + item.session + "_run-" + item.run
+
+            for (key in item.keySet()){
+                if(item[key] == 'todo'){
+                    error "Error ~ Please look at your tractoflow_bids_struct.json " +
+                    "in Read_BIDS folder.\nPlease fix todo fields and give " +
+                    "this file in input using --bids_config option instead " +
```
Comment (AlexVCaron) [inline main.nf]: instead of

File: main.nf [hunk 3]
```diff
@@ -157,42 +161,135 @@ workflow.onComplete {
...
+                    "using --bids."
+                }
+                else if (item[key] == 'error_readout'){
+                    error "Error ~ Please look at your tractoflow_bids_struct.json " +
+                    "in Read_BIDS folder.\nPlease fix error_readout fields. "+
+                    "This error indicate that readout time looks wrong.\n"+
+                    "Please correct the value or remove the subject in the json and " +
+                    "give the updated file in input using --bids_config option instead " +
```
Comment (AlexVCaron) [inline main.nf]: instead of

File: main.nf [hunk 4]
```diff
@@ -157,42 +161,135 @@ workflow.onComplete {
...
+
+if (number_subj_for_null_check == 0){
+    error "Error ~ No subjects found. Please check the naming convention or your BIDS folder."
+}
+
+number_subj_for_compare.count()
+    .concat(number_rev_b0_for_compare)
+    .toList()
+    .subscribe{a, b -> if (a != b && b > 0) 
+    error "Error ~ Some subjects have a reversed phase encoded b=0 and others no.\n" +
```
Comment (AlexVCaron) [inline main.nf]: others don't

File: USAGE [hunk 5]
```diff
@@ -25,11 +25,17 @@ DESCRIPTION
                                             ├── rev_b0.nii.gz  (optional)
                                             └── t1.nii.gz
 
-    --dti_shells                                Shells selected to compute the dti metrics (generally b < 1200).
-                                                Please write them between quotes e.g. (--dti_shells "0 300").
+    --bids=/path/to/[bids]              BIDS folder following the BIDS convention.
 
-    --fodf_shells                               Shells selected to compute the fodf metrics (generally b > 700).
-                                                Please write them between quotes e.g. (--fodf_shells "0 1000").
+    --bids_config                       BIDS configuration file in json pre-generated by TractoFlow(Read_BIDS process).
```
Comment (AlexVCaron) [inline USAGE:30]: Could we auto-detect the presence of this config file ? Then if its present, fixed and --bids is used, we just load it, no need to change the command lines.

Comment (arnaudbore) [reply USAGE:30]: I find this suggestion dangerous. I won't go there or maybe later if it's a need from the community.

**Comment by jchoude** [comment]
When there is an issue in the BIDS formatting (in my case, missing a `dataset_description` file), the `ReadBids` process fails. It is retried 3 times, but when everything fails, NF is just hung. It will "continue working" indefinitely, instead of stopping.

Probably need to change the error strategy for that process.

File: main.nf (hunk 4 above)
Comment (GuillaumeTh) [reply main.nf]: Done

File: main.nf (hunk 3 above)
Comment (GuillaumeTh) [reply main.nf]: Done

File: main.nf (hunk 2 above)
Comment (GuillaumeTh) [reply main.nf]: Done

**Comment by GuillaumeTh** [comment]
@jchoude The new errorStrategy is 3 retry and after crash.

**Comment by jchoude** [comment]
Didn't really look at the code, but tested on different datasets, and issues were fixed. Most of them were in the scilpy reader.

**Review by jchoude — APPROVED** [review]

File: main.nf [hunk 6]
```diff
@@ -157,42 +161,136 @@ workflow.onComplete {
     log.info "Execution duration: $workflow.duration"
 }
 
-if (params.root){
+if (params.root && (!params.bids || !params.bids_config)){
```
Comment (AlexVCaron) [inline main.nf]:
```suggestion
if (params.root && !(params.bids && params.bids_config)){
```

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
One little change could be made for one condition, anything else seems good to me !

**Pull request merged**